llama_index/indices/base.py

```diff
@@ -340,30 +342,65 @@ def as_query_engine(self, **kwargs: Any) -> BaseQueryEngine:
         return RetrieverQueryEngine.from_args(**kwargs)
 
     def as_chat_engine(
-        self, chat_mode: ChatMode = ChatMode.CONDENSE_QUESTION, **kwargs: Any
+        self, chat_mode: ChatMode = ChatMode.BEST, **kwargs: Any
     ) -> BaseChatEngine:
+        query_engine = self.as_query_engine(**kwargs)
+        if "service_context" not in kwargs:
+            kwargs["service_context"] = self._service_context
+
+        # resolve chat mode
+        if chat_mode == ChatMode.BEST:
+            # get LLM
+            service_context = cast(ServiceContext, kwargs["service_context"])
+            llm = service_context.llm
+
+            if isinstance(llm, OpenAI) and is_function_calling_model(llm.model):
+                chat_mode = ChatMode.OPENAI
+            else:
+                chat_mode = ChatMode.REACT
+
         if chat_mode == ChatMode.CONDENSE_QUESTION:
             # NOTE: lazy import
             from llama_index.chat_engine import CondenseQuestionChatEngine
 
-            query_engine = self.as_query_engine(**kwargs)
-            if "service_context" not in kwargs:
-                kwargs["service_context"] = self._service_context
             return CondenseQuestionChatEngine.from_defaults(
                 query_engine=query_engine,
                 **kwargs,
             )
-        elif chat_mode == ChatMode.REACT:
+        elif chat_mode in [ChatMode.REACT, ChatMode.OPENAI]:
             # NOTE: lazy import
-            from llama_index.chat_engine import ReActChatEngine
-
-            query_engine = self.as_query_engine(**kwargs)
-            if "service_context" not in kwargs:
-                kwargs["service_context"] = self._service_context
-            return ReActChatEngine.from_query_engine(
-                query_engine=query_engine,
-                **kwargs,
+            from llama_index.agent import OpenAIAgent, ReActAgent
+            from llama_index.tools.query_engine import QueryEngineTool
+
+            # build query engine tool
+            name = kwargs.pop("name", None)
+            description = kwargs.pop("description", None)
```

Comment:
are these kwargs documented?

Reply:
they are not, in hind sight this is way too hidden anyways. will remove.